File: typescript/src/ships/internal/loadout-state.ts
/** State-boundary helpers for the mutable loadout facade. @internal */

import { deepFreeze } from '../../internal/deep-freeze.js';
import type { OutfittingModule } from '../modules.js';
import type { LoadoutModule } from '../slef.js';
import type { BuildSlot } from '../slots.js';

/**
 * How a fitted module's `Item` is named when a catalogue lookup rejects it as a wrong
 * type. A build assembled through `setModule` carries a symbol the guard there checked;
 * one read from a `Loadout` event carries whatever the event held, so the failure has to
 * say which field of the build it came from.
 */
export const FITTED_ITEM = 'ShipLoadout: fitted module Item';

/** Whether a journal key names a known cosmetic or hull-geometry entry. */
export function isNonOutfittingSlot(slotKey: string): boolean {
    return /^(paintjob|shipname\d+|shipid\d+|bobble\d+|decal\d+|shipkit.+|weaponcolour|enginecolour|vesselvoice|shipcockpit|stringlights)$/i.test(
        slotKey,
    );
}

/** Whether a module is the zero-mass, zero-price cargo hatch built into every hull. */
export function isBuiltInHullModule(module: LoadoutModule): boolean {
    return (
        module.Slot.toLowerCase() === 'cargohatch' &&
        module.Item.toLowerCase().startsWith('modularcargobaydoor')
    );
}

/** Detach a journal module from caller-owned or returned mutable objects. */
export function cloneLoadoutModule(module: LoadoutModule): LoadoutModule {
    return {
        Slot: module.Slot,
        Item: module.Item,
        ...(module.On === undefined ? {} : { On: module.On }),
        ...(module.Priority === undefined ? {} : { Priority: module.Priority }),
        ...(module.Health === undefined ? {} : { Health: module.Health }),
        ...(module.Value === undefined ? {} : { Value: module.Value }),
        ...(module.Engineering === undefined
            ? {}
            : {
                  Engineering: {
                      BlueprintName: module.Engineering.BlueprintName,
                      Level: module.Engineering.Level,
                      Quality: module.Engineering.Quality,
                      ...(module.Engineering.ExperimentalEffect === undefined
                          ? {}
                          : { ExperimentalEffect: module.Engineering.ExperimentalEffect }),
                      ...(module.Engineering.ExperimentalEffect_Localised === undefined
                          ? {}
                          : {
                                ExperimentalEffect_Localised:
                                    module.Engineering.ExperimentalEffect_Localised,
                            }),
                      ...(module.Engineering.Modifiers === undefined
                          ? {}
                          : {
                                Modifiers: module.Engineering.Modifiers.map((modifier) => ({
                                    ...modifier,
                                })),
                            }),
                  },
              }),
    };
}

/**
 * The key a slot-keyed map actually holds for a mount, or `null` when it holds none.
 *
 * @remarks
 * A build's own spelling is authoritative and is never rewritten — Frontier writes
 * `FrameShiftDrive` where a SLEF producer may write `frameshiftdrive`, and both name the
 * same mount — so every read and every mutation resolves the caller's key through here
 * first. The own-key hit is a fast path: only a miss pays for the scan. It never has to
 * break a tie, because a build cannot hold two keys differing only in case (`fromLoadout`
 * throws on one, and every edit writes through this).
 */
export function matchingKeyIn(keyed: ReadonlyMap<string, unknown>, slotKey: string): string | null {
    if (keyed.has(slotKey)) return slotKey;
    const wanted = slotKey.toLowerCase();
    for (const key of keyed.keys()) {
        if (key.toLowerCase() === wanted) return key;
    }
    return null;
}

/** Order values by a hull's slots, retaining source order for slots outside the layout. */
export function orderBySlotLayout<T>(
    values: readonly T[],
    layout: readonly BuildSlot[] | null | undefined,
    slotOf: (value: T) => string,
): T[] {
    if (!layout) return [...values];
    const order = new Map(layout.map((slot, index) => [slot.key.toLowerCase(), index] as const));
    return [...values].sort(
        (left, right) =>
            (order.get(slotOf(left).toLowerCase()) ?? Number.MAX_SAFE_INTEGER) -
            (order.get(slotOf(right).toLowerCase()) ?? Number.MAX_SAFE_INTEGER),
    );
}

/** Snapshot caller-supplied stats so later caller mutation cannot alter the build. */
export function cloneModuleStats(module: OutfittingModule): OutfittingModule {
    try {
        return deepFreeze(structuredClone(module));
    } catch (error) {
        if (!(error instanceof Error && error.name === 'DataCloneError')) throw error;
        return deepFreeze({
            ...module,
            ...(module.restrictedToShips === undefined
                ? {}
                : { restrictedToShips: [...module.restrictedToShips] }),
            ...(module.limitIncrease === undefined
                ? {}
                : { limitIncrease: { ...module.limitIncrease } }),
            ...(module.damageDistribution === undefined
                ? {}
                : { damageDistribution: { ...module.damageDistribution } }),
            ...(module.damageComponents === undefined
                ? {}
                : {
                      damageComponents: {
                          ...module.damageComponents,
                          ...(module.damageComponents.unclassified === undefined
                              ? {}
                              : { unclassified: [...module.damageComponents.unclassified] }),
                      },
                  }),
            ...(module.projectileRange === undefined
                ? {}
                : { projectileRange: { ...module.projectileRange } }),
        });
    }
}

File: dotnet/src/EliteDangerousAlmanac/Ships/Internal/LoadoutState.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

namespace EliteDangerousAlmanac.Ships.Internal;

/// <summary>The state-boundary rules a build's own store keeps.</summary>
internal static class LoadoutState
{
    /// <summary>
    /// The catalogued cargo hatch every hull's mount resolves to.
    /// </summary>
    /// <remarks>
    /// A hull-specific hatch symbol is not catalogued separately, so this one record carries
    /// the article's stats for all of them.
    /// </remarks>
    internal const string BuiltInHullSymbol = "ModularCargoBayDoor";

    private static readonly Regex NonOutfittingKey = new(
        @"^(paintjob|shipname\d+|shipid\d+|bobble\d+|decal\d+|shipkit.+|weaponcolour|enginecolour|vesselvoice|shipcockpit|stringlights)$",
        RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);

    /// <summary>Whether a journal key names a known cosmetic or hull-geometry entry.</summary>
    internal static bool IsNonOutfittingSlot(string slotKey) => NonOutfittingKey.IsMatch(slotKey);

    /// <summary>Whether a journal key names the cargo-hatch mount.</summary>
    internal static bool IsCargoHatchSlot(string slotKey) =>
        string.Equals(slotKey, "CargoHatch", StringComparison.OrdinalIgnoreCase);

    /// <summary>Whether a symbol names the cargo hatch built into every hull.</summary>
    /// <remarks>
    /// This tests a prefix rather than the whole symbol, because some hull families name their
    /// own hatch — the Fer-de-Lance family's <c>ModularCargoBayDoorFDL</c> — for the one
    /// article the catalogue carries. It is the article's identity, so the fit rules and the
    /// import path read the same answer from it.
    /// </remarks>
    internal static bool IsBuiltInHullSymbol(string symbol) =>
        symbol.StartsWith(BuiltInHullSymbol, StringComparison.OrdinalIgnoreCase);

    /// <summary>Whether a fitted module is the weightless, free hatch every hull carries.</summary>
    internal static bool IsBuiltInHullModule(LoadoutModule module) =>
        IsCargoHatchSlot(module.Slot) && IsBuiltInHullSymbol(module.Item);

    /// <summary>
    /// The key a build actually holds for a mount, or <see langword="null"/> when it holds none.
    /// </summary>
    /// <remarks>
    /// A build's own spelling is authoritative and is never rewritten. Frontier writes
    /// <c>FrameShiftDrive</c> where a SLEF producer may write <c>frameshiftdrive</c>, and both
    /// name the same mount, so every read and every edit resolves a caller's key through here
    /// first. There is never a tie to break, because a build cannot hold two keys that differ
    /// only in case.
    /// </remarks>
    internal static string? MatchingKey(IReadOnlyList<LoadoutModule> fitted, string slotKey)
    {
        int index = IndexOf(fitted, slotKey);
        return index < 0 ? null : fitted[index].Slot;
    }

    /// <summary>
    /// The build's existing spelling for a mount, or its own casing convention for a new one.
    /// </summary>
    /// <remarks>
    /// A build read from an export that lower-cases every key keeps doing so when an edit fills
    /// an empty mount, so one build never mixes two conventions.
    /// </remarks>
    internal static string OwnKey(IReadOnlyList<LoadoutModule> fitted, string canonicalKey)
    {
        string? held = MatchingKey(fitted, canonicalKey);
        if (held is not null) return held;
        if (fitted.Count == 0) return canonicalKey;

        foreach (LoadoutModule module in fitted)
        {
            if (!IsLowerCase(module.Slot)) return canonicalKey;
        }

        return canonicalKey.ToLowerInvariant();
    }

    private static bool IsLowerCase(string key)
    {
        foreach (char letter in key)
        {
            if (char.IsUpper(letter)) return false;
        }

        return true;
    }

    /// <summary>The index of one mount in a build's store, or -1 when it holds none.</summary>
    internal static int IndexOf(IReadOnlyList<LoadoutModule> fitted, string slotKey)
    {
        for (int index = 0; index < fitted.Count; index++)
        {
            if (string.Equals(fitted[index].Slot, slotKey, StringComparison.OrdinalIgnoreCase)) return index;
        }

        return -1;
    }

    /// <summary>
    /// Puts values in the hull's own mount order, keeping the source order of anything the
    /// layout does not name.
    /// </summary>
    internal static List<T> OrderByLayout<T>(
        IReadOnlyList<T> values,
        IReadOnlyList<BuildSlot> layout,
        Func<T, string> slotOf)
    {
        Dictionary<string, int> order = new(layout.Count, StringComparer.OrdinalIgnoreCase);
        for (int index = 0; index < layout.Count; index++) order[layout[index].Key] = index;

        return values.OrderBy(value =>
            order.TryGetValue(slotOf(value), out int rank) ? rank : int.MaxValue).ToList();
    }
}
